### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO) and Post Link Optimization (PLO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [gfx-rs/wgpu#4692] [Ideas]
Hi!

Recently I checked Profile-Guided Optimization (PGO) improvements on multiple projects. The results are [here](https://github.com/zamazan4ik/awesome-pgo/). PGO results for LLVM-related tooling are [here](https://github.com/llvm/llvm-project/issues/63486). According to the tests, PGO usually helps with the compiler and compiler-like workloads (like static analysis or code formatters) - e.g. Clang [gets](https://llvm.org/docs/HowToBuildWithPGO.html#introduction) +20% compilation speed with PGO. Since this, I think trying to optimize Wgpu tools like `naga-cli`, would be a good idea. I already did some PGO benchmarks and want to share my results here.

## Test environment

I performed tests on two configurations: Linux and macOS.

Linux:

* Fedora 39, KDE, Wayland
* Linux kernel 6.5.11
* AMD Ryzen 9 5900x
* 48 Gib RAM
* SSD Samsung 980 Pro 2 Tib
* Compiler - Rustc 1.73
* GPU: Nvidia 3080 Ti
* Wgpu version: the latest for now from the `trunk` branch on commit `611471c4bc8b1adab4d7d96fd2ba8bae82dccb28`

macOS:

* OS: macOS 13.6
* CPU: Macbook M1 Pro 6+2
* Wgpu version: the latest for now from the `trunk` branch on commit `611471c4bc8b1adab4d7d96fd2ba8bae82dccb28`

## Benchmark

For benchmark purposes, I run `cargo bench --workspace --no-fail-fast -p naga --bench criterion`. PGO training phase was done on the same scenario with `cargo pgo bench -- --workspace --no-fail-fast -p naga --bench criterion`. PGO optimization is done with [cargo-pgo](https://github.com/Kobzol/cargo-pgo).

Benchmarks are done on the same machine, with the same background "noise". The benchmark was performed multiple times - the results are reproducible.

## Results

I got the following results (in `cargo bench` format):

Linux:

* PGO-optimized compared to Release bench: https://gist.github.com/zamazan4ik/ef03ce9cee39839202f716098841bc9f
* (just for reference) PGO-instrumented compared to Release bench: https://gist.github.com/zamazan4ik/df96b8ddc081a354b63cce69710ed476

macOS:

* PGO-optimized compared to Release bench: https://gist.github.com/zamazan4ik/fe21310f2638cc7ce9355d5ec0a719d2
* (just for reference) PGO-instrumented compared to Release bench: https://gist.github.com/zamazan4ik/3958de6255af93ec68b7526ea125ab8c

At least in the scenario above PGO helps with achieving better performance.

One benchmark does not work on my setup but right I have no idea how to fix it.

## Further steps

I can suggest the following action points:

* Perform more PGO benchmarks on Wgpu tools. If it shows improvements - add a note to the documentation about possible improvements in Wgpu tools' performance with PGO.
* Providing an easier way (e.g. a build option) to build scripts with PGO can be helpful for the end-users and maintainers since they will be able to optimize Wgpu tools according to their workloads.
* Optimize pre-built binaries with PGO (if it's possible to prepare a generic enough profile).

Testing Post-Link Optimization techniques (like [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md)) would be interesting too (Clang and Rustc already use BOLT as an addition to PGO) but I recommend starting from the usual PGO.

Here are some examples of how PGO optimization is already integrated into other projects:

* Rustc: a CI [script](https://github.com/rust-lang/rust/blob/master/src/ci/stage-build.py) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
  - A [part](https://github.com/gcc-mirror/gcc/blob/4832767db7897be6fb5cbc44f079482c90cb95a6/configure#L7818) in a "wonderful" `configure` script 
* Clang: [Docs](https://llvm.org/docs/HowToBuildWithPGO.html) 
* Python: 
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* [Foot](https://codeberg.org/dnkl/foot): [Scripts](https://codeberg.org/dnkl/foot/src/branch/master/pgo)
* Windows Terminal: [GitHub PR](https://github.com/microsoft/terminal/pull/10071)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)
* file.d: [GitHub PR](https://github.com/ozontech/file.d/pull/469)
* OceanBase: [CMake flag](https://github.com/oceanbase/oceanbase/blob/master/cmake/Env.cmake#L55)

I am not sure how performance is critical for Wgpu now. If it isn't the biggest priority right now - could be a nice feature to get in the future.

**Comment by cwfitzgerald** [comment]
This is all very lovely work and a good reference for anyone interested in PGO themselves - but I don't think there's anything actionable that we can do here.

We're fundamentally a library that people use in their own programs, so they are in charge of the compile settings used. We do have a few binaries (wgpu-info, naga-cli, etc) but we don't actually publish binaries for them as they're just developer tools. Everyone who uses them uses them through `cargo install` which recompiles the binary on the system. We are again not in control of building.

It's good to know that PGO does help naga (it makes sense, it's very branch heavy).

**Reply by zamazan4ik** [reply]
> This is all very lovely work and a good reference for anyone interested in PGO themselves - but I don't think there's anything actionable that we can do here.

Even if Wgpu is a library-first project, I think it's possible to do several things.

We can add a note somewhere in the Wgpu documentation about possible PGO effects (at least based on the benchmarks above). This information could be useful for:

* For the people who use wgpu as a library. So they can estimate somehow benefits by enabling PGO for Wgpu and decide to recompile their project with PGO.
* For the people who use and/or maintain Wgpu binaries (`naga-cli` is probably the most performance-critical thing but I am not sure). Based on PGO benefits numbers, they can decide to recompile their local binaries with PGO to get an additional performance boost (if it's critical for them).

Here I have some PGO-related documentation examples in several projects:

* ClickHouse: https://clickhouse.com/docs/en/operations/optimizing-performance/profile-guided-optimization
* Databend: https://databend.rs/doc/contributing/pgo
* Vector: https://vector.dev/docs/administration/tuning/pgo/
* Nebula: https://docs.nebula-graph.io/3.5.0/8.service-tuning/enable_autofdo_for_nebulagraph/
* GCC: Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
* Clang:
  - https://llvm.org/docs/HowToBuildWithPGO.html
  - https://llvm.org/docs/AdvancedBuilds.html
* tsv-utils: https://github.com/eBay/tsv-utils/blob/master/docs/BuildingWithLTO.md (yet, it's actually about PGO too)

If we are talking about binary optimization, even if wgpu does not distribute prebuilt binaries, we can make building Wgpu tools with PGO easier via some predefined build tweaks. The idea is quite simple:

* Prepare some generic enough training workload (if it's possible to do so)
* Write some Cargo/CMake/make/whatever integration to build a tool like `naga-cli` with the profile from the workload above. Something like `make pgooptimize`. Not sure about how it could be integrated into the Cargo build workflow - perhaps something with `build.rs`.
* A user locally runs simply `make pgooptimize` and gets PGO optimized version

In this case, for the users would be easier to get PGO optimized program version.

I have several examples of how it looks like in other projects:

* Rustc: a CI [tool](https://github.com/rust-lang/rust/tree/master/src/tools/opt-dist) for the multi-stage build
* GCC:
  - Official [docs](https://gcc.gnu.org/install/build.html), section "Building with profile feedback" (even AutoFDO build is supported)
  - A [part](https://github.com/gcc-mirror/gcc/blob/master/configure#L7896) in a "wonderful" `configure` script.
* Clang: [Docs](https://llvm.org/docs/HowToBuildWithPGO.html)
* Python:
  - CPython: [README](https://github.com/python/cpython#profile-guided-optimization)
  - Pyston: [README](https://github.com/pyston/pyston#building)
* Go: [Bash script](https://github.com/golang/go/blob/master/src/cmd/compile/profile.sh)
* Swift: [CMake script](https://github.com/apple/swift/blob/main/CMakeLists.txt#L364)
* V8: [Bazel flag](https://github.com/v8/v8/blob/main/BUILD.gn#L184)
* ChakraCore: [Scripts](https://github.com/chakra-core/ChakraCore/tree/master/Build/scripts/pgo)
* Chromium: [Script](https://chromium.googlesource.com/chromium/src/build/config/+/refs/heads/main/compiler/pgo/BUILD.gn)
* Firefox: [Docs](https://firefox-source-docs.mozilla.org/build/buildsystem/pgo.html)
   - Thunderbird has PGO support too
* PHP - [Makefile command](https://github.com/php/php-src/blob/master/build/Makefile.global#L138) and old Centminmod [scripts](https://github.com/centminmod/php_pgo_training_scripts)
* MySQL: [CMake script](https://github.com/mysql/mysql-server/blob/8.0/cmake/fprofile.cmake)
* YugabyteDB: [GitHub commit](https://github.com/yugabyte/yugabyte-db/commit/34cb791ed9d3d5f8ae9a9b9e9181a46485e1981d)
* FoundationDB: [Script](https://github.com/apple/foundationdb/blob/1a6114a66f3de508c0cf0a45f72f3687ba05750c/contrib/generate_profile.sh)
* Zstd: [Makefile](https://github.com/facebook/zstd/blob/dev/programs/Makefile#L232)
* [Foot](https://codeberg.org/dnkl/foot): [Scripts](https://codeberg.org/dnkl/foot/src/branch/master/pgo)
* Windows Terminal: [GitHub PR](https://github.com/microsoft/terminal/pull/10071)
* Pydantic-core: [GitHub PR](https://github.com/pydantic/pydantic-core/pull/741)
* file.d: [GitHub PR](https://github.com/ozontech/file.d/pull/469)
* OceanBase: [CMake flag](https://github.com/oceanbase/oceanbase/blob/master/cmake/Env.cmake#L55)
* ISPC: [CMake scipts](https://github.com/ispc/ispc/tree/main/superbuild)
* NodeJS: [Configure script](https://github.com/nodejs/node/commit/9be15559cc0bfe506d9cdfba4ad0f4beacf5ce17)
* Android Open Source Project (AOSP):
  - [Official documentation](https://source.android.com/docs/core/perf/pgo)
  - Committed PGO profiles: [repository](https://android.googlesource.com/toolchain/pgo-profiles/+/refs/heads/main)
* DMD: [Custom build rule](https://github.com/dlang/dmd/blob/master/compiler/src/build.d#L553)
* LDC: [GitHub action](https://github.com/ldc-developers/ldc/blob/master/.github/actions/2a-build-pgo/action.yml)
* tsv-utils: [Makefile](https://github.com/eBay/tsv-utils/blob/master/makefile#L56)
* Erlang OTP: [Makefile](https://github.com/erlang/otp/blob/master/Makefile.in#L484)
* Clingo (PGO enabled only in Spack): [Package recipe](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/clingo-bootstrap/package.py#L96)


Which suggestions are worth it from the "benefits/implementation cost" point of view - is all up to you. IMO putting a note into the documentation is a cheap enough change, providing a build infra PGO integration could be a more time-consuming task to complete.
